File: leetcode/easy/414.Third Maximum Number/js/thirdMaximumNumber.js
"use strict";

/**
 * @param {number[]} nums
 * @return {number} third distinct maximum number 
 * @author AlexSoto
 */

// 1. sort array burble method descending order
// 2. maxNumber = -999_999_999, countMaxNumber = 0
// 3. while countMaxNumber === 3...
// 3.1 for i equal 0, while i < nums length...
// 3.1.1 if i equal 0 and nums[i] > maxNumber....
// 3.1.1.1 countMaxNumber += 1
// 3.1.1.2 maxNumber = nums[i]
// 3.1.1 else if nums[i] < maxNumber...
// 3.1.1.1 countMaxNumber += 1
// 3.1.1.2 maxNumber = nums[i]
// 3.1.1 else countMaxNumber === 3:
// 3.1.1.1 return maxNumber
// 4. return maxNumber

var thirdMax = function(nums) {
    console.log(nums);
    let isItSorted = false;
    while (isItSorted === false) {
        isItSorted = true;
        for (let i = 0; i < nums.length - 1; i++) {
            if (nums[i] < nums[i + 1]) {
                let temp = nums[i];
                nums[i] = nums[i + 1];
                nums[i + 1] = temp;
                isItSorted = false;
            };
        };
    };
    console.log(nums);
    let maxNumber = nums[0];
    let countMaxNumber = 1;
    for (let i = 0; i < nums.length; i++) {
        if (nums[i] < maxNumber) {
            maxNumber = nums[i];
            countMaxNumber++;
        };
        if (countMaxNumber === 3) {
            return maxNumber;
        };
    };
    if (countMaxNumber != 3) {
        return nums[0];
    };
};

// Testing
let nums0 = [3,2,1];
let nums1 = [1,2];
let nums2 = [2,2,3,1];
let nums3 = [1,2,2,5,3,5];
console.log(thirdMax(nums3));

File: leetcode/easy/414.Third Maximum Number/python/thirdMaximumNumber.py
"""Third Maximum Number"""
def third_maximum_number(nums):
    set_nums = set(nums)
    if len(set_nums) <= 2:
        return max(nums)
    else:
        list_nums = list(set_nums)
        list_nums.sort() 
        list_nums = list_nums[-3:]
    return min(list_nums)
        

data1 = [-4,-5,-3,-2,-1]
print(third_maximum_number(data1))

'''
1.
input: integer array
output: integer number which is the result of figure out what is the distinct maximum number

2. 
edge cases: if lenght of nums <=2 return the maximum
key: sort the array
cast it to set
conditional statment to know if the set greater or equal to 3
return the min of that set

3. 
conditional statement if to know if nums <= 2 if so
    then return max(nums)
else:
    cast nums list to set
    sort set
    create a list from the last three elements
    return min(list)

4. happy coding!
'''
# My Approach
def third_maximum_number(nums):
    nums.sort()
    tries = 1
    greatest_num = nums[-1]
    index = len(nums) - 2
    while index >= 0:
        if nums[index] < greatest_num:
            greatest_num = nums[index]
            tries += 1
        if tries == 3:
            break
        if index == 0:
            return nums[-1]
        index -= 1
    return greatest_num

# Testing
data_test_0 = [3,2,1]
data_test_1 = [2,1]
data_test_2 = [2,2,3,1]
print(third_maximum_number(data_test_1))

'''Complexity Analysis
Time Complexity: O(NlogN)
Space Complexity: O(1)'''

# Testing
data_test_0 = [1,2,-2147483648]
print(third_maximum_number(data_test_0))